src/base/static/components/input-form/input-form.js

```diff
@@ -0,0 +1,776 @@
+      attrs[fieldName] = attrs[fieldName].replace(
+        /\<img.*?name=\"(.*?)\".*?\>/g,
+        "{{#rich_text_image $1}}"
+      );
+    });
+
+    if (this.richTextImages.length > 0) {
+      this.richTextImages.forEach(richTextImage => {
+        model.attachmentCollection.add(richTextImage);
+      });
+    }
+
+    if (coverImages.length > 0) {
+      coverImages.forEach(coverImage => {
+        model.attachmentCollection.add(coverImage);
+      });
+    }
+
+    // TODO: is this still necessary?
+    // Util.setStickyFields(  
+    //   attrs,
+    //   Shareabouts.Config.survey.items,
+    //   Shareabouts.Config.place.items,
+    // );
+
+    // TODO: case when url-title field is left blank: fall back to regular urls
+
+    attrs.geometry = this.geometry;
+    if (this.geometryStyle) {
+      attrs.style = this.geometryStyle;
+    }
+
+    // Fire pre-save hook.
+    // The pre-save hook allows flavors to attach arbitrary data to the attrs
+    // object before submission to the database.
+    if (customHooks && customHooks.preSave) {
+      attrs = hooks[customHooks.preSave](attrs);
+    }
+
+    model.save(attrs, {
+      success: (response) => {
+        Util.log("USER", "new-place", "successfully-add-place");
+        this.setState({ "formIsSubmitting": false });
+
+        // Fire post-save hook.
+        // The post-save hook allows flavors to hijack the default
+        // route-to-detail-view behavior.
+        if (customHooks && customHooks.postSave) {
+          customHooks.postSave(response, model, this.defaultPostSave.bind(this));
+        } else {
+          this.defaultPostSave(model);
+        }
+      },
+      error: (error) => {
+        Util.log("USER", "new-place", "fail-to-add-place");
+      },
+      wait: true,
+    });
+  }
+
+  defaultPostSave(model) {
+    this.reset();
+    this.props.router.navigate(Util.getUrl(model), { trigger: true });
+  }
+
+  getVisibleFormCategories() {
+    return this.props.placeConfig.place_detail
+      // Filter out categories that shouldn't display on the form at all.
+      .filter(categoryConfig => categoryConfig.includeOnForm)
+      // Filter out admin_only categories that shouldn't be shown given the
+      // current user's credentials.
+      .filter(categoryConfig => {
+        if (categoryConfig.admin_only &&
+            !Util.getAdminStatus(categoryConfig.dataset, categoryConfig.admin_groups)) {
+          return false;
+        }
+        return true;
+      });
+  }
+
+  fieldIsInvalid(fieldName) {
+    return (
+      this.state.formValidationErrors.length > 0 
+      && this.requiredFields.includes(fieldName)
+      && !this.state.fieldValues[fieldName]
+    );
+  }
+
+  render() {
+    const { categoryMenuIsCollapsed, categoryMenuIsHidden, formIsSubmitting,
+            formValidationErrors, selectedCategory } = this.state;
+    const { className, hideCenterPoint, hideSpotlightMask, placeConfig, 
```

Comment:
nit: Is there a reason why we are redefining new variables for these functions? Many of them are only being used once. 

I think these would be a little easier to read if we left the `props.` prefix in front, making it clear that these functions are coming from our props.

Reply:
The pattern I tried to follow was that if we destructure `state` or `props`, we destructure every attribute that we use in a given method. There are a bunch of places where we have single-use destructured variables--should the criteria in general be that we only destructure attributes if they're used more than once?